Deprecate execute()

1ucian0 opened the pull request:
## Summary

The execute() function is as old as the SDK and it was always a high-level combinant function to say "here, a circuit, do your thing". That is exactly what Qiskit Primitives are about. Time to deprecated it.

Closes #7892 
Closes #7640

mtreinish commented on qiskit/execute_function.py:
```diff
@@ -35,6 +36,13 @@ def _log_submission_time(start_time, end_time):
 
+@deprecate_func(
+    additional_msg="This function combines ``transpile`` and ``backend.run``, which is covered "
+    "by ``Sampler`` primitive. Alternatively, you can also run :func:`.transpile` "
```
Do you want to add links for the sampler and primitives too?

1ucian0 replied:
in this context, it is Sampler "as a primitive". Linking it to the reference implementation is an option, but not the only one. Do you suggest to explain that in text?

mtreinish replied:
You cold do:
```suggestion
    "by ``Sampler`` :mod:`~qiskit.primitives`. Alternatively, you can also run :func:`.transpile` "
```
Which would link to this page: https://qiskit.org/documentation/apidoc/primitives.html

That being said `BackendSampler` is the closest to what execute does: https://qiskit.org/documentation/stubs/qiskit.primitives.BackendSampler.html#qiskit.primitives.BackendSampler

mtreinish commented on releasenotes/notes/deprecate-execute-67a5c68b93722057.yaml:
```diff
@@ -0,0 +1,20 @@
+    Qiskit's :func:`~.execute_function.execute` function is deprecated.
+    This function served as a high-level wrapper around transpiling
+    a circuit with some transpile options and running it on a backend
+    with some run options. To do the same thing, you can explicitly
```
Do you want to call out the sampler here too? The `BackendSampler` class is fairly straightforward mapping to what `execute()` does (although imo `execute()` or` transpile()` followed by `backend.run()` are much easier to work with).

1ucian0 replied:
I added an example in 0367d407c0347343e0f0bfdfa8cfc3428d4799c1

jakelishman commented:
(A minor note: CI's likely to be busy today, so if we're able to try and keep the number of commits we submit to CI down by doing more testing locally, it'll help keep the rc1 release more to time.)

1ucian0 committed:
squash and change of base

1ucian0 committed:
Update qiskit/execute_function.py

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

1ucian0 committed:
test/visual/mpl/graph/test_graph_matplotlib_drawer.py

1ucian0 committed:
buggy bug

1ucian0 committed:
transpile

1ucian0 committed:
Merge branch 'stable/0.46' into deprecate/execute/1

jakelishman commented on test/python/basicaer/test_basicaer_integration.py:
```diff
@@ -33,7 +32,7 @@ def setUp(self):
         self._qc1.measure(qr[0], cr[0])
         self.backend = BasicAer.get_backend("qasm_simulator")
-        self._result1 = execute(self._qc1, self.backend).result()
+        self._result1 = self.backend.run(self._qc1).result()
```
Not that I imagine it will matter here, but technically this is a behavioural change due to the missing `transpile`.  Same comment for most changes in the next two files.

jakelishman commented on test/python/basicaer/test_multi_registers_convention.py:
```diff
@@ -39,17 +39,17 @@ def test_circuit_multi(self):
 
         backend_sim = BasicAer.get_backend("unitary_simulator")
-        result = execute(circ, backend_sim, seed_transpiler=3438).result()
+        result = backend_sim.run(circ, seed_transpiler=3438).result()
```
These `seed_transpiler` kwargs look spurious now that the `transpile` call has been elided.

1ucian0 replied:
oops, my bad. Removing it in 05dd6869a88ad51e42e8ae682e82f6fa06e837fd

jakelishman commented on test/python/circuit/test_controlled_gate.py:
```diff
@@ -488,7 +489,7 @@ def test_multi_controlled_u1_matrix(self, num_controls):
 
             backend = BasicAer.get_backend("unitary_simulator")
-            simulated = execute(qc, backend).result().get_unitary(qc)
+            simulated = backend.run(transpile(qc, backend)).result().get_unitary(qc)
```
I wouldn't mind just changing these into `simulated = Operator(qc)`, but this is also fine.

1ucian0 replied:
done in 6ef382aab412f744a330b36fbc8eda6bc623082c

jakelishman commented on test/python/providers/test_fake_backends.py:
```diff
@@ -102,9 +101,8 @@ def setUpClass(cls):
+            transpile(self.circuit, backend),
             optimization_level=optimization_level,
             seed_simulator=42,
             seed_transpiler=42,
```
This, and the other test in this file, set `seed_transpiler` which ideally needs moving into the `transpile` call.

1ucian0 replied:
indeed. Done in 2fa258009c9243a6342125a7de4f459f3534b89b

jakelishman reviewed (commented):
Thanks for this, Luciano.  A few more notes for consideration:

- in `examples/python`, there's still uses of `execute`.  I don't know if we still even want to keep that folder, but if so, we probably ought to change them to at least be `backend.run`/`transpile` if not `Sampler`.
- `test/randomized` still has a couple of uses of `execute`
- the docstring of `qiskit.quantum_info.analysis.distance.hellinger_fidelity` still uses `execute` in an example

I noticed there's a few places in the tests where `execute` was replaced only with `backend.run` (no `transpile`), typically for `BasicAer` backends.  I suspect that's probably a fine change and won't affect anything, just pointing it out.  There's a few of those changes that have left a spurious `seed_transpiler` argument in `backend.run`, though, which it looks like BasicAer just ignores.  We should probably tidy those up.

1ucian0 committed:
Merge branch 'stable/0.46' into deprecate/execute/1

1ucian0 committed:
examples

1ucian0 committed:
test/randomized

1ucian0 committed:
hellinger_fidelity

1ucian0 committed:
clean seed_transpiler

1ucian0 commented:
Good catches!

> * in `examples/python`, there's still uses of `execute`.  I don't know if we still even want to keep that folder, but if so, we probably ought to change them to at least be `backend.run`/`transpile` if not `Sampler`.

Fixed in 39535b9201b52bcc253d4d9ded68d41af39b9d64

> * `test/randomized` still has a couple of uses of `execute`

Fixed in 1bea14bc6eff01c0e89cc05ad98b3f6a665db11b

> * the docstring of `qiskit.quantum_info.analysis.distance.hellinger_fidelity` still uses `execute` in an example

Fixed in 39535b9201b52bcc253d4d9ded68d41af39b9d64
 
> I noticed there's a few places in the tests where `execute` was replaced only with `backend.run` (no `transpile`), typically for `BasicAer` backends. I suspect that's probably a fine change and won't affect anything, just pointing it out.

Yes. It was intentional for the cases where it was possible, as a way to minimize the changes. 

> There's a few of those changes that have left a spurious `seed_transpiler` argument in `backend.run`, though, which it looks like BasicAer just ignores. We should probably tidy those up.

done in 05dd6869a88ad51e42e8ae682e82f6fa06e837fd

1ucian0 committed:
seed_transpiler in test_fake_backends

1ucian0 committed:
Operator

jakelishman committed:
Fix pulse documentation

jakelishman committed:
Fix overzealous deprecation skip

jakelishman committed:
Remove other 'execute' uses in tests

jakelishman reviewed (dismissed):
I pushed up a couple more commits just to squash a couple more places I spotted, or places where we using a warning skip when `Operator(qc)` was just as easy.

I found one place that was skipping warnings for `bind_parameters` with a `module='test\.python\.circuit\.test_parameters'` (or something) that was causing an `execute` to get missed, but that's actually a _good_ think - it shows the `stacklevel` was set correctly, otherwise it wouldn't have been skipped by that!  (I fixed the skip by adding a regex match for messages mentioning the intended target.)

There's a couple of places that still use `execute` in tests along with other deprecated or dead code that's being tidied up elsewhere, so I've just left them.  Let's get this into the merge queue.

jakelishman committed:
Fix unused import
